### PR TITLE
Change `failure_status` into a map

### LINKED ISSUE #10403: List of failure statuses? Would help with testing, and help users

### PR DESCRIPTION
**Pull request by ghost-not-in-the-shell**
Explain your changes:
This PR changes the `failure_status` in transaction_logic.ml to be a map from `public_key` to `Transaction_status.Failure.t list`. The purpose is to collect all the failures when applying a snapp.

Explain how you tested your changes:

Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [ ] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [x] Serialized types are in stable-versioned modules
- [x] Does this close issues? List them

* Closes #10403

### CODE REVIEW

**Review by mrmr1993 — COMMENTED** [review]
If multiple parties use the same public key, won't this collapse them all into the same list (particularly so with tokens)? It feels like we might be better associating the failures with the index into the party list, so that it's clear which parties have failed with which errors

File: src/lib/mina_base/transaction_logic.ml [hunk 1]
```diff
@@ -1365,18 +1365,17 @@ module Make (L : Ledger_intf) : S with type ledger := L.t = struct
...
-            (* Raise a more useful error message if we have a failure
-               description.
-            *)
-            Error.raise
-              (Error.of_string @@ Transaction_status.Failure.to_string failure)
-        | _ ->
-            assert b
+        if (not b) && not (Hashtbl.is_empty failure_status) then
+          (* Raise a more useful error message if we have a failure
+                description. *)
```
Comment (psteckler) [inline src/lib/mina_base/transaction_logic.ml:1375]: nit: indentation of comment

File: src/lib/mina_base/parties_logic.ml [hunk 2]
```diff
@@ -33,6 +33,8 @@ module type Bool_intf = sig
 
   val all : t list -> t
 
+  type single_failure_status
```
Comment (psteckler) [inline src/lib/mina_base/parties_logic.ml:36]: could we make this type `failure_status`, and the other one `failure_status_tbl`?

File: src/lib/mina_base/transaction_status.ml [hunk 3]
```diff
@@ -47,6 +47,28 @@ module Failure = struct
     end
   end]
 
+  module Table = struct
```
Comment (mrmr1993) [inline src/lib/mina_base/transaction_status.ml:50]: Nit: `Table` normally refers to the `Core` hash table type. Could you rename this? (`Collection`, perhaps?)

**Review by mrmr1993 — APPROVED** [review]

**Pull request merged**